### ISSUE TITLE
Ordered list of inline Lua code blocks

### ISSUE THREAD

**Issue by mhite** [envoyproxy/envoy#10996]
Is it possible to attach an order list of inline Lua code blocks into a listener HTTP filter configuration?

For example, I have one block of code that checks for a header of `X-Forwarded-Proto: http` and issues a redirect to https if discovered. I would want this particular Lua code block to run first and attach as its own Envoy configuration entity. For example:

```
function envoy_on_request(request_handle)
  local headers = request_handle:headers()
  local xfp = headers:get('x-forwarded-proto')
  if xfp == "http" then
    local path = headers:get(':path')
    local authority = headers:get(':authority')
    local location = "https://" .. authority .. path
    request_handle:respond({[":status"] = "308", ["Location"] = location}, "308 Permanent Redirect")
  end
end
```

Similarly, there are occasions on some of my proxy servers where I'd like to also chain the following code to also run:

```
function envoy_on_response(response_handle)
  function hasFrameAncestors(rh)
    s = rh:headers():get("Content-Security-Policy");
    delimiter = ";";
    defined = false;
    for match in (s..delimiter):gmatch("(.-)"..delimiter) do
      match = match:gsub("%s+", "");
      if match:sub(1, 15)=="frame-ancestors" then
        return true;
      end
    end
    return false;
  end
  if not response_handle:headers():get("Content-Security-Policy") then
    csp = "frame-ancestors none;";
    response_handle:headers():add("Content-Security-Policy", csp);
  elseif response_handle:headers():get("Content-Security-Policy") then
    if not hasFrameAncestors(response_handle) then
      csp = response_handle:headers():get("Content-Security-Policy");
      csp = csp .. ";frame-ancestors none;";
      response_handle:headers():replace("Content-Security-Policy", csp);
    end
  end
  if not response_handle:headers():get("X-Frame-Options") then
    response_handle:headers():add("X-Frame-Options", "deny");
  end
  if not response_handle:headers():get("X-XSS-Protection") then
    response_handle:headers():add("X-XSS-Protection", "1; mode=block");
  end
  if not response_handle:headers():get("X-Content-Type-Options") then
    response_handle:headers():add("X-Content-Type-Options", "nosniff");
  end
  if not response_handle:headers():get("Referrer-Policy") then
    response_handle:headers():add("Referrer-Policy", "no-referrer");
  end
  if not response_handle:headers():get("X-Download-Options") then
    response_handle:headers():add("X-Download-Options", "noopen");
  end
  if not response_handle:headers():get("X-DNS-Prefetch-Control") then
    response_handle:headers():add("X-DNS-Prefetch-Control", "off");
  end
  if not response_handle:headers():get("Feature-Policy") then
    response_handle:headers():add("Feature-Policy",
                                  "camera 'none';"..
                                  "microphone 'none';"..
                                  "geolocation 'none';"..
                                  "encrypted-media 'none';"..
                                  "payment 'none';"..
                                  "speaker 'none';"..
                                  "usb 'none';");
  end
  if response_handle:headers():get("X-Powered-By") then
    response_handle:headers():remove("X-Powered-By");
  end
end
```

My naive reading of the API seems to suggest there is a single `envoy.filters.http.lua` filter that can be set through configuration on `envoy.http_connection_manager` list of `httpFilters`. This would seem to suggest if one wanted to create a precedence among various Lua code blocks, they would need to arranged as a single code block in a single filter config API entity. 

Perhaps I'm superimposing expectations/world-views that were set elsewhere. Specifically, in the world of F5, an iRule is tcl code that lets you reference connection/stream events and "hook" into them. You can load multiple code blocks as separate configuration entities, order them, and the proxy takes care of merging all the like event names in the code and runs them merged in your configured order.

My first look at the Envoy lua filter was that maybe it might let me do this as it seems to provide an event API (ie. `envoy_on_response`, `envoy_on_request`). I'm operating Envoy through Istio, so there is a level of indirection (`EnvoyFilter`) that sometimes gets in the way of me effectively troubleshooting (and sometimes understanding the underbelly of) some of this stuff. My first experiments seem to suggest that it is not possible to attach multiple Lua filters to a `envoy.http_connection_manager` filter.

For anyone who has deeper knowledge of this, would you mind helping to set me straight?

If there is no such functionality to merge "events", would this be useful for others? Does it inherently not fit the envoy model?

Thanks!

**Comment by mhite** [comment]
This appears to (I think?) touch on the multiple filter idea of the same type:
 https://github.com/envoyproxy/envoy/issues/9358

**Comment by zuercher** [comment]
I don't have time to check at the moment, but the way it might be possible (from the Envoy perspective) is configure multiple LUA filters with distinct names using `typed_config` (with an `@type` field) to identify themselves as LUA filters. Assuming that works at the Envoy level, there's a separate question of whether a config with that shape can be rendered by Istio.

**Comment by snowp** [comment]
It's perfectly possible to list the same filter multiple times, Envoy doesn't really care - I've personally ran Envoy with 4-5 lua filters back to back. 

The requirement around distinct names just comes into play if you want to do things like provide different per_filter_config to different filters where you might want to provide different per filter configuration to the different filters.

**Comment by mhite** [comment]
@snowp - Gotcha! I had originally tried using a `typed_config` but the version of Istio we are currently running (1.4.6) was rejecting it:

```
kubectl apply -f ./en-istio-helm/securityHeaders-envoyfilter.yaml
Error from server: error when creating "./en-istio-helm/securityHeaders-envoyfilter.yaml": admission webhook "pilot.validation.istio.io" denied the request: configuration is invalid: Envoy filter: unknown message type "envoy.config.filter.http.lua.v2.Lua"
```
I suspect a `typed_config` for Lua becomes available in 1.5, because I see it directly referenced in the 1.5 documentation:

https://istio.io/docs/reference/config/networking/envoy-filter/

(Just search for `"type.googleapis.com/envoy.config.filter.http.lua.v2.Lua"` -- you'll see it referenced)

Whereas in 1.4 documentation, the examples don't use `typed_config` for Lua but instead an untyped-config with name `envoy.lua`:

https://archive.istio.io/v1.4/docs/reference/config/networking/envoy-filter/

Does this mean the type is inferred from the name in the non-typed config case?

If so, then I'm guessing that I need to run an Istio version which supports the Lua `typed_config` type so that I can use unique names and arrange these in an appropriate order in the filter list.

Am I on the right track?
